src/codegen/compile.ml

```diff
@@ -4529,39 +4603,18 @@ module Serialization = struct
     let name = "@serialize_go<" ^ typ_hash t ^ ">" in
     Func.share_code3 env name (("x", I32Type), ("data_buffer", I32Type), ("ref_buffer", I32Type)) [I32Type; I32Type]
     (fun env get_x get_data_buf get_ref_buf ->
-      let set_data_buf = G.setter_for get_data_buf in
       let set_ref_buf = G.setter_for get_ref_buf in
 
       (* Some combinators for writing values *)
 
-      let advance_data_buf =
-        get_data_buf ^^ G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^ set_data_buf in
-
-      let write_word code =
-        let (set_word, get_word) = new_local env "word" in
-        code ^^ set_word ^^
-        I32Leb.compile_store_to_data_buf_unsigned env get_word get_data_buf ^^
-        advance_data_buf
-      in
-
-      let write_word32 code =
-        get_data_buf ^^ code ^^
-        G.i (Store {ty = I32Type; align = 0; offset = 0l; sz = None}) ^^
-        compile_unboxed_const Heap.word_size ^^ advance_data_buf
-      in
-
-      let write_byte code =
-        get_data_buf ^^ code ^^
-        G.i (Store {ty = I32Type; align = 0; offset = 0l; sz = Some Wasm.Types.Pack8}) ^^
-        compile_unboxed_const 1l ^^ advance_data_buf
-      in
+      let write_word, write_word32, write_byte, write_blob, write_text, write_unsigned, write_signed = Strm.(write_word_leb env get_data_buf, write_word_32 env get_data_buf, write_byte env get_data_buf, write_blob env get_data_buf, write_text env get_data_buf, write_unsigned env get_data_buf, write_signed env get_data_buf) in
```

Comment:
Use `List.map`

Reply:
Cant we use the module system to make this prettier, @crusso? (map won't work if the result types are different)

Reply:
But here the types are the same, so a homogeneous `List` will do. This looks like a _very_ specific use case for a module system.

Also I did it this way to minimise the changes relative to the original code. Sounded better to me to apply boring arguments here then changing a lot of call sites.

Reply:
Yeah, I would actually use a local functor for this, parameterizing Stream by env and get_data_buf, and instantiating once - at least that way there is less pausibility of mismatching the entries. You could use open to dump the bindings but I think projecting from the module would actually be clearer...

Reply:
But the databuf value isnt constant, is it? in different functions they are different

Reply:
All I want is
``` ocaml
      let    [write_word;     write_word32;  write_byte; write_blob; write_text; write_unsigned;   write_signed] =
        Strm.[write_word_leb; write_word_32; write_byte; write_blob; write_text; write_bignum_leb; write_bignum_sleb]
        |> List.map (fun f -> f env get_data_buf) in
```
But I get a `Warning 8 [partial-match]: this pattern-matching is not exhaustive.` Damn! `map` is structure-preserving...

Reply:
Yep, it isn't worth the complication. 1d32004dc just cuts the Gordian knot.